pkg/server/handler.go

```diff
@@ -0,0 +1,248 @@
+	isProcessing bool
+	ctx          context.Context
+	cancelFunc   context.CancelFunc
+
+	// for unit test
+	HandlerEngine
+}
+
+type HandlerEngine interface {
+	GetSize(url string) (int64, error)
+	Download(ctx context.Context, url, filePath string, updater util.ProgressUpdater) (written int64, err error)
+	ReceiverLaunch(ctx context.Context, port string, filePath string, syncFileOps sparserest.SyncFileOperations) error
+	SenderLaunch(localPath string, remote string, timeout int, directIO bool) error
+	UploadServerLaunch(ctx context.Context, port string, directory string, syncFileOps sparserest.SyncFileOperations, sizeUpdater uploadserver.SizeUpdater) error
+}
+
+func (d *BackingImageHandler) GetSize(url string) (int64, error) {
+	return d.HandlerEngine.GetSize(url)
+}
+
+func (d *BackingImageHandler) InitProcessing() error {
+	d.Lock()
+	defer d.Unlock()
+	if d.ctx != nil && d.cancelFunc != nil {
+		return fmt.Errorf("handler is already initialized")
+	}
+	if d.ctx != nil || d.cancelFunc != nil {
+		d.cancelProcessingWithoutLock()
+		return fmt.Errorf("handler is not initialized correctly")
+	}
+	d.ctx, d.cancelFunc = context.WithCancel(context.Background())
+	return nil
+}
+
+func (d *BackingImageHandler) Pull(url, filePath string, updater util.ProgressUpdater) (written int64, err error) {
+	d.Lock()
+	if d.isProcessing {
+		d.Unlock()
+		return 0, fmt.Errorf("handler is already processing")
+	}
+	if d.ctx == nil || d.cancelFunc == nil {
+		d.cancelProcessingWithoutLock()
+		d.Unlock()
+		return 0, fmt.Errorf("BUG: handler is not initialized correctly or already cancelled before actual pulling")
+	}
+	ctx := d.ctx
+	d.isProcessing = true
+	d.Unlock()
+
+	defer func() {
+		d.Cancel()
+	}()
+	return d.HandlerEngine.Download(ctx, url, filePath, updater)
+}
+
+func (d *BackingImageHandler) Receive(port string, filePath string, syncFileOps sparserest.SyncFileOperations) error {
+	d.Lock()
+	if d.isProcessing {
+		d.Unlock()
+		return fmt.Errorf("handler is already processing")
+	}
+	if d.ctx == nil || d.cancelFunc == nil {
+		d.cancelProcessingWithoutLock()
+		d.Unlock()
+		return fmt.Errorf("BUG: handler is not initialized correctly or already cancelled before actual receiving")
+	}
+	ctx := d.ctx
+	d.isProcessing = true
+	d.Unlock()
+
+	defer func() {
+		d.Cancel()
+	}()
+
+	if err := d.HandlerEngine.ReceiverLaunch(ctx, port, filePath, syncFileOps); err != nil && err != http.ErrServerClosed {
+		return err
+	}
+	return nil
+}
+
+// Send should fail once the receiver is closed or the timeout is reached.
+func (d *BackingImageHandler) Send(filePath string, address string) error {
+	if d.isProcessingStart() {
+		return fmt.Errorf("handler cannot send files when the processing is not finished")
+	}
+
+	return d.HandlerEngine.SenderLaunch(filePath, address, types.FileSyncTimeout, false)
+}
+
+func (d *BackingImageHandler) Upload(port string, directory string, syncFileOps sparserest.SyncFileOperations, sizeUpdater uploadserver.SizeUpdater) error {
```

Comment:
Not sure if there is a better design for this API. Right now inputs `syncFileOps sparserest.SyncFileOperations` and `sizeUpdater uploadserver.SizeUpdater` are actually the same BackingImage object